benchmarks/data_generator/README.md

```diff
@@ -0,0 +1,124 @@
+{"timestamp": 0, "input_length": 7234, "output_length": 794, "hash_ids": [0, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41]}
+{"timestamp": 0, "input_length": 2287, "output_length": 316, "hash_ids": [0, 42, 43, 44, 45]}
+```
+Note that each new hash id is the next consecutive integer after the last one used. To generate these increasing hash ids from a list of texts, we provide the `texts_to_hashes` function in `hasher.py`.
+
+## Synthesizer
+
+The Synthesizer goes a step further:
+It builds a prefix tree from the original trace file, extracts prefix statistics, and generates a new synthetic dataset based on these statistics.
+You can control various aspects of the synthetic data generation with tunable knobs, such as request rate, context/prompt length multipliers, and the number of tree copies.
+
+This is useful for generating large, realistic synthetic traces for benchmarking or simulation, while preserving the structural properties of the original dataset.
+
+### How to run
+
+```bash
+datagen synthesize --input-file <path_to_trace.jsonl> --num-requests <N> [other options...]
+```
+
+**Options:**
+- `--input-file`: Path to the input trace file (default: `mooncake_trace.jsonl`)
+- `--num-requests`: Number of requests to synthesize (default: 100000)
+- `--speedup-ratio`: Factor to speed up request intervals. It effectively divides the synthetic timestamps by this value (default: 1)
+- `--depth-multiplier`: Multiplier for prefix lengths (default: 1.0)
+- `--width-multiplier`: Number of times to replicate the core radix tree (default: 1)
+- `--prompt-len-multiplier`: Multiplier for leaf path lengths (default: 1.0, use <1 for shorter prompts)
+- `--max-isl`: Maximum input sequence length to include in output (default: None, no filtering)
+- `--block-size`: Block size for prefilling and decoding (default: 512)
+- `--output-file`: Path to the output file (default: auto-generated from input file and options)
+
+### Example
+
+Say we only have these hash lists:
+
+```
+[0, 1, 2, (3)]
+[0, 1]
+[0, 1, 2]
+[0, (4), (5)]
+```
+
+First, we identify the "core prefix nodes" as [0, 1, 2] since they are visited more than once. The nodes [3, 4, 5] would be considered "user prompts" as they only appear once (noted in brackets).
+
+If we set the `depth-multiplier` to 2, then the core prefix branches will be stretched, effectively giving:
+
+```
+[0, 1, 2, 3, 4, 5, 6, (7)]
+[0, 1, 2, 3]
+[0, 1, 2, 3, 4]
+[0, 1, (8), (9)]
+```
+
+
+Note that the "prompt branches" are not stretched by `depth-multiplier`. They can be separately modified by applying `prompt-len-multiplier`.
+
+Now, if we set `width-multiplier` to 2, then each row will have a 50 percent chance of being incremented by a large integer, so that they will be effectively separated into a new radix tree, which matches the statistics of the original one, but having completely different roots.
```

Comment:
this one seems very odd w.r.t to the other terms here.

multiplier - would be multiplying something and here it is used as a percentage?  Maybe this should be 'prefix-root-multiplier'